libs/api/ebpf_api.cpp

```diff
@@ -1464,26 +1469,31 @@ _delete_ebpf_object(_In_opt_ _Post_invalid_ ebpf_object_t* object) noexcept
     }
 }
 
-static void
-_remove_ebpf_object_from_globals(_In_ const ebpf_object_t* object) noexcept
+_Requires_lock_not_held_(_ebpf_state_mutex) static void _remove_ebpf_object_from_globals(
+    _In_ const ebpf_object_t* object) noexcept
 {
+    std::unique_lock lock(_ebpf_state_mutex);
     EBPF_LOG_ENTRY();
     ebpf_assert(object);
     auto it = std::find(_ebpf_objects.begin(), _ebpf_objects.end(), object);
     ebpf_assert(it != _ebpf_objects.end());
     _ebpf_objects.erase(it);
 }
 
-static void
-_clean_up_ebpf_objects() noexcept
+_Requires_lock_not_held_(_ebpf_state_mutex) static void _clean_up_ebpf_objects() noexcept
 {
     EBPF_LOG_ENTRY();
-    for (auto& object : _ebpf_objects) {
+    std::unique_lock lock(_ebpf_state_mutex);
+    auto objects = std::move(_ebpf_objects);
+    _ebpf_objects.resize(0);
+
+    lock.unlock();
+
+    for (auto& object : objects) {
         _delete_ebpf_object(object);
     }
 
-    _ebpf_objects.resize(0);
-
+    lock.lock();
     ebpf_assert(_ebpf_programs.size() == 0);
     ebpf_assert(_ebpf_maps.size() == 0);
```

Comment:
IMO, using scopes instead of explicitly calling `unlock()` and `lock()` again would make the code more readable and less fragile. It would require you to declare `object` outside/before the scope, so you couldn't use `auto` (which I like too, so 😢 ). Finally, do you really _need_ the asserts at the end, or could they just be above when you hold the lock initially?